internal/hcsoci/credentials.go

```diff
@@ -0,0 +1,136 @@
+// +build windows
+
+package hcsoci
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/Microsoft/hcsshim/internal/hcs"
+	"github.com/Microsoft/hcsshim/internal/log"
+	hcsschema "github.com/Microsoft/hcsshim/internal/schema2"
+	"github.com/sirupsen/logrus"
+)
+
+// This file holds the necessary structs and functions for adding and removing Container
+// Credential Guard instances (shortened to CCG normally) for V2 HCS schema
+// containers. Container Credential Guard is in HCS's own words "The solution to
+// allowing windows containers to have access to domain credentials for the
+// applications running in their corresponding guest." It essentially acts as
+// a way to temporarily Active Directory join a given container with a Group
+// Managed Service Account (GMSA for short) credential specification.
+// CCG will launch a process in the host that will act as a middleman for the
+// credential passthrough logic. The guest is then configured through registry
+// keys to have access to the process in the host.
+// A CCG instance needs to be created through various HCS calls and then added to
+// the V2 schema container document before being sent to HCS. For V1 HCS schema containers
+// setting up instances manually is not needed, the GMSA credential specification
+// simply needs to be present in the V1 container document.
+
+// CCGInstance stores the id used when creating a ccg instance. Used when
+// closing a container to be able to release the instance.
+type CCGInstance struct {
+	// ID of container that instance belongs to.
+	id string
+}
+
+// Release calls into hcs to remove the ccg instance. These do not get cleaned up automatically
+// they MUST be explicitly removed with a call to ModifyServiceSettings. The instances will persist
+// unless vmcompute.exe exits or they are removed manually as done here.
+func (instance *CCGInstance) Release(ctx context.Context) error {
+	if err := removeCredentialGuard(ctx, instance.id); err != nil {
+		log.G(ctx).WithFields(logrus.Fields{
```

Comment:
Don't think we need to log here, the returned error should be sufficient, right?

Reply:
Just following suit with the other Release methods that log if there's an err

Reply:
@kevpar would you prefer this be removed?

Reply:
It's not a big deal, but good to cut down on extraneous logging where we can. Thinking about this more we have a potential issue where if one of the other resources fails to close, we could leak the CCG instance and not ever clean it up. This isn't an issue for most of our resources since they'll go away when the vmwp dies, but the CCG instance doesn't have its lifetime tied to anything else.

Reply:
I think we should change the code in `resources.go` so that if a resource fails to `Release`, we log the error there but continue on with cleaning up the remaining resources, and then return an error at the end. That will prevent the CCG instance from being leaked if an earlier resource fails to clean up.

However, that doesn't help us if something else happens, like the shim crashing. In that case, I think we would leak the CCG instance, as nothing else knows about it to clean it up.

I think we will need to add code somewhere that enumerates all CCG instances and cleans up any that shouldn't be running anymore. However, where to add that code, and how it should decide that, is non-obvious.

Reply:
@kevpar The CCG instances are added last in the slice so they should always be cleaned up first so the first half shouldn't be a worry. I don't know if I like these little ordering edge cases but thats another discussion. 

So a summary of some changes proposed and please give feedback
1. Change the Release funcs to stop logging an error and instead return the err we were logging and log this in ReleaseResources. 
2. If any of the resources fail to release in ReleaseResources, don't return the error immediately but wait till after the loop to return an error so we can continue on with the others. 
3. On shim startup after we create an instance and query for it if there's more than the instance with the ID we care about we could remove them. This is what I'd like feedback on the most.

Reply:
I know it's currently added at the end of the resources, but we shouldn't rely on that in general. :)

For #3, that won't work if there is more than one concurrent container using GMSA. I feel like we might need code in containerd to clean this up properly.

Reply:
No I agree, that's why I'd rather implement 1 and 2 I was just saying as of now this wouldn't occur. Is there anything we can do in this PR specifically?

Reply:
I think it's okay to get this in, file a new bug to track the remaining issue, and ensure we mention to our partners and in the release notes that this is a potential issue.

Reply:
Cool, I'll file it and do 1 & 2